rubix-spi/src/test/java/com/qubole/rubix/spi/TestBookKeeperFactory.java

```diff
@@ -0,0 +1,195 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License. See accompanying LICENSE file.
+ */
+package com.qubole.rubix.spi;
+
+import com.qubole.rubix.spi.thrift.BookKeeperService;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.thrift.TException;
+import org.apache.thrift.server.TServer;
+import org.apache.thrift.server.TThreadPoolServer;
+import org.apache.thrift.transport.TServerSocket;
+import org.apache.thrift.transport.TServerTransport;
+import org.apache.thrift.transport.TTransportException;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.testng.Assert.assertTrue;
+
+public class TestBookKeeperFactory
+{
+  private static final Log log = LogFactory.getLog(TestBookKeeperFactory.class);
+
+  private static final int NO_DELAY = 0;
+  private static final int WAIT_FOR_BKS_START = 200;
+
+  private final Configuration conf = new Configuration();
+  private BookKeeperFactory bookKeeperFactory;
+
+  @BeforeMethod
+  public void setUp()
+  {
+    bookKeeperFactory = new BookKeeperFactory();
+  }
+
+  @AfterMethod
+  public void tearDown()
+  {
+    conf.clear();
+  }
+
+  @Test
+  public void testCreateBookKeeperClient_ableToConnect() throws TException, InterruptedException
+  {
+    final int socketTimeout = 500;
+    final int connectTimeout = 500;
+
+    MockBookKeeperServer server = startMockServer(true, NO_DELAY, NO_DELAY);
+
+    final RetryingBookkeeperClient client = createTestBookKeeperClient(socketTimeout, connectTimeout);
+    assertTrue(client.isBookKeeperAlive());
+
+    stopMockServer(server);
+  }
+
+  @Test(expectedExceptions = TTransportException.class)
+  public void testCreateBookKeeperClient_unableToConnect_connectTimeout() throws TException, InterruptedException
+  {
+    final int startDelay = 1000;
+    final int socketTimeout = 500;
+    final int connectTimeout = 500;
+
+    startMockServer(false, startDelay, NO_DELAY);
+
+    createTestBookKeeperClient(socketTimeout, connectTimeout); // should throw expected exception due to connect timeout
+  }
+
+  @Test(expectedExceptions = TTransportException.class)
+  public void testCreateBookKeeperClient_unableToConnect_socketTimeout() throws TException, InterruptedException
+  {
+    final int aliveCallDelay = 1000;
+    final int socketTimeout = 500;
+    final int connectTimeout = 500;
+
+    startMockServer(true, NO_DELAY, aliveCallDelay);
+
+    final RetryingBookkeeperClient client = createTestBookKeeperClient(socketTimeout, connectTimeout);
+    client.isBookKeeperAlive(); // should throw expected exception due to socket timeout
+  }
+
```

Comment:
Can we add a test where we are able to connect although there is a delay in the server ( delay is less than the socket time out). In that case we are not going to get any exception

Reply:
@abhishekdas99 Added tests to cover more cases